File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque


def solution(bridge_length, weight, truck_weights):
    time = 0
    sum = 0

    bridge = deque([0]*bridge_length)
    wait = deque(truck_weights)

    while wait:
        sum -= bridge[0]
        bridge.popleft()
        time += 1

        if ((sum + wait[0]) > weight):
            bridge.append(0)

        else:
            truck = wait.popleft()
            sum += truck
            bridge.append(truck)

    return time + bridge_length

File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int sum = 0;
        
        Queue<Integer> wait = new LinkedList<>();
        for (int w : truck_weights) {
            wait.add(w);
        }
        
        Queue<Integer> br = new LinkedList<>();
        for (int i = 0; i < bridge_length; i++) {
            br.add(0);
        }
        
        while (!br.isEmpty()) {
            br.poll();
            answer++;
            sum = br.stream().mapToInt(i -> i).sum();
            if (!wait.isEmpty()) {
                if (wait.peek() + sum <= weight) {
                    br.add(wait.poll());
                } else {
                    br.add(0);
                }
            }
        }
        return answer;
    }
}